lahja/endpoint.py

```diff
@@ -61,36 +65,177 @@ def from_name(cls, name: str, base_path: Optional[pathlib.Path]=None) -> 'Connec
+
+        async with self._drain_lock:
+            # Use a lock to serialize drain() calls. Circumvents this bug:
+            # https://bugs.python.org/issue29930
+            await self.writer.drain()
+
+    async def read_message(self) -> Message:
+        raw_size = await self.reader.readexactly(4)
+        size = int.from_bytes(raw_size, 'little')
+        message = await self.reader.readexactly(size)
+        obj = pickle.loads(message)
+        assert isinstance(obj, Message)
+        return obj
+
+    @property
+    def more_to_read(self) -> bool:
+        return not self.reader.at_eof()
 
 
-class ProxyEndpointConnector(BaseProxy):  # type: ignore # TypeShed is missing BaseProxy
+class InboundConnection:
     """
-    Proxy that connects to an :class:`~lahja.endpoint.EndpointConnector`
+    An Endpoint might have several `InboundConnection`s, each of them represents an
+    Endpoint which has connected to the given Endpoint and is attempting to send it
+    messages.
     """
+    def __init__(self, conn: Connection, new_msg_func: Callable[[Broadcast], None]) -> None:
+        self.conn = conn
+        self.new_msg_func = new_msg_func
+
+        self.logger = logging.getLogger('lahja.endpoint.InboundConnection')
+
+        self._notify_lock = asyncio.Lock()
+        self._received_response = asyncio.Condition()
+
+    async def run(self) -> None:
+        try:
+            while self.conn.more_to_read:
+                message = await self.conn.read_message()
+
+                if isinstance(message, Broadcast):
+                    self.new_msg_func(message)
+                elif isinstance(message, SubscriptionsAck):
+                    async with self._received_response:
+                        self._received_response.notify_all()
+                else:
+                    self.logger.error(f'received unexpected message: {message}')
+        except asyncio.IncompleteReadError:
+            # the remote endpoint has disconnected
+            pass
+
+    async def notify_subscriptions_updated(self,
+                                           subscriptions: Set[Type[BaseEvent]],
+                                           block: bool = True) -> None:
+        """
+        Alert the Endpoint which has connected to us that our subscription set has
+        changed. If `block` is True then this function will block until the remote
+        endpoint has acknowledged the new subscription set. If `block` is False then this
+        function will return immediately after the send finishes.
+        """
+        async with self._notify_lock:
+            # The lock ensures only one coroutine can notify this endpoint at any one time
+            # and that no replies are accidentally received by the wrong coroutines.
+            async with self._received_response:
+                await self.conn.send_message(SubscriptionsUpdated(subscriptions, block))
+                if block:
+                    await self._received_response.wait()
```

Comment:
I kind of understand this but only because I've spent some brain-space in the past trying to understand the use case for `asyncio.Condition`.  Can you maybe create a large-ish comment that tries to lay out exactly what is happening here and why.  This interplay between synchronization primatives seems appropriate but also hard for a newcomer to understand.